### PR TITLE
Add the option to not generate maybeMaps and maybeWhen

### PR DESCRIPTION
**Pull request by Lyokone**
In order to solve #562
This PR aims at adding options to make the generation of maybeMap and maybeWhen optional.

Not sure if I added the feature the correct way, any feedback is greatly appreciated if I can help the PR to land. I know it's something we would use in our next projects.

### CODE REVIEW

**Comment by rrousselGit** [comment]
This is pretty good!

Could you add some documentation about this in the readme? Including the ability to disable it globally using `build.yaml`
Once this is done, I'll merge it

**Comment by Lyokone** [comment]
Hi!
I've added the documentation to the maybewhen and maybemap section of the Readme.

File: packages/freezed_annotation/lib/freezed_annotation.dart [hunk 1]
```diff
@@ -129,6 +131,40 @@ class Freezed {
...
+  /// ```
+  ///
+  /// The generated code will not have the following method:
+  ///
+  /// ```dart
+  /// void main() {
+  ///   print(Union.first().maybeMap(orElse: () => null)); // Error
+  /// }
+  /// ```
+  final bool shouldGenerateMaybeMap;
```
Comment (rrousselGit) [inline packages/freezed_annotation/lib/freezed_annotation.dart:150]: Looking back at this, the "shouldGenerate" is probably redundant. I think `@Freezed(maybeMap: false)` is enough

Comment (Lyokone) [reply packages/freezed_annotation/lib/freezed_annotation.dart]: I've changed the front facing `Freezed` class but kept `shouldGenerateMaybeMap` in `Concrete` and `Abstract` to match `shouldGenerateJson`. Let me know if you prefer to change it everywhere :)

File: packages/freezed_annotation/lib/freezed_annotation.dart [hunk 2]
```diff
@@ -11,6 +11,8 @@ class Freezed {
     this.unionKey,
     this.unionValueCase = FreezedUnionCase.none,
     this.fallbackUnion,
+    this.maybeMap = true,
+    this.maybeWhen = true,
```
Comment (rrousselGit) [inline packages/freezed_annotation/lib/freezed_annotation.dart:15]:
```suggestion
    this.maybeMap,
    this.maybeWhen,
```

File: packages/freezed_annotation/lib/freezed_annotation.dart [hunk 3]
```diff
@@ -129,6 +131,40 @@ class Freezed {
...
+  /// ```
+  ///
+  /// The generated code will not have the following method:
+  ///
+  /// ```dart
+  /// void main() {
+  ///   print(Union.first().maybeMap(orElse: () => null)); // Error
+  /// }
+  /// ```
+  final bool maybeMap;
```
Comment (rrousselGit) [inline packages/freezed_annotation/lib/freezed_annotation.dart:150]:
```suggestion
  final bool? maybeMap;
```

File: packages/freezed_annotation/lib/freezed_annotation.dart [hunk 4]
```diff
@@ -129,6 +131,40 @@ class Freezed {
...
+  /// ```
+  ///
+  /// The generated code will not have the following method:
+  ///
+  /// ```dart
+  /// void main() {
+  ///   print(Union.first().maybeWhen(orElse: () => null)); // Error
+  /// }
+  /// ```
+  final bool maybeWhen;
```
Comment (rrousselGit) [inline packages/freezed_annotation/lib/freezed_annotation.dart:167]:
```suggestion
  final bool? maybeWhen;
```

**Comment by rrousselGit** [comment]
LGTM thanks!

**Comment by smiLLe** [comment]
@Lyokone  Was there a specific reason to not include the OrNull methods?

**Pull request merged**